### PR TITLE
Rename dig.Param to dig.In

### PR DESCRIPTION
**Pull request by glibsm**
Step 1 of #80 

I'm ok to wait and change the base branch to master after we merge all of dev in. I will, however, get a few more PRs out to get more of the functionality in for the named instances.

Blocked on #73 and will need rebase afterwards

### CODE REVIEW

**Comment by glibsm** [comment]
Rename drops coverage? Cool story, gocov

File: types.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// in the graph, memebers of the struct are inserted into the graph.
+//
+// SturctTags on those memebers control their behavior. For example,
+//
+// 		type Input struct {
+//			dig.In
+//
+//      S *Something
+//			T *Thingy `optional:"true"`
+//    }
```
Comment (abhinav) [inline types.go]: comment indentation is off

File: types.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package dig
+
+// In is an embeddable object that signals to dig that the struct
+// should be treated differently. Instead of itself becoming an object
+// in the graph, memebers of the struct are inserted into the graph.
+//
+// SturctTags on those memebers control their behavior. For example,
```
Comment (abhinav) [inline types.go]: nit: s/SturctTags/Tags/

File: types.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+//
+// 		type Input struct {
+//			dig.In
+//
+//      S *Something
+//			T *Thingy `optional:"true"`
+//    }
+//
+// TODO better usage docs
+type In struct{}
```
Comment (abhinav) [inline types.go]: nit: move TODO out of the actual docs

File: types.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+var _ digInObject = In{}
+
+// Param is the only instance of parameterObject.
+func (In) digInObject() {}
```
Comment (abhinav) [inline types.go]: outdated comment

File: dig.go [hunk 5]
```diff
@@ -30,8 +30,8 @@ import (
 var (
 	_noValue             reflect.Value
 	_errType             = reflect.TypeOf((*error)(nil)).Elem()
-	_parameterObjectType = reflect.TypeOf((*parameterObject)(nil)).Elem()
-	_paramType           = reflect.TypeOf(Param{})
+	_parameterObjectType = reflect.TypeOf((*digInObject)(nil)).Elem()
+	_paramType           = reflect.TypeOf(In{})
```
Comment (abhinav) [inline dig.go]: optional: rename to _inType and _inObjectType?

File: types.go (hunk 3 above)
Comment (glibsm) [inline types.go]: didn't you write this comment? :)

**Comment by breerly** [comment]
It's not gocov - coveralls has their own bs check

File: types.go (hunk 3 above)
Comment (abhinav) [inline types.go]: True. I noticed the fault and moved it out in a recent PR.

**Pull request merged**